packages/context/src/context.ts

```diff
@@ -64,65 +112,167 @@ export class Context {
     return this.registry.has(key);
   }
 
+  /**
+   * Iterate all contexts by breadth first traversal of the graph following
+   * `parents`. For example, with the following context graph:
+   * - reqCtx -> [serverCtx, connectorCtx]
+   * - serverCtx -> [appCtx]
+   * - connectorCtx -> [appCtx]
+   * `reqCtx.contexts()` returns an iterator of `[reqCtx, serverCtx,
+   * connectorCtx, appCtx]`.
+   */
+  protected *contexts(): IterableIterator<Context> {
+    const visited: Set<Context> = new Set();
+    const queue: Context[] = [];
+    // Enqueue the current context
+    queue.push(this);
+    while (queue.length) {
+      // Dequeue the head context
+      const c = queue.shift()!;
+      // Check if the context has been added
+      if (!visited.has(c)) {
+        visited.add(c);
+        yield c;
+        // Enqueue the parent contexts
+        queue.push(...c.parents);
+      }
+    }
+  }
+
+  /**
+   * Visit all contexts in the graph. It first calls the `visitor` function to
+   * produce a `result` against the current context. If the `result.done` flag
+   * is set to `true`, the method returns `result.value`. Otherwise, it iterates
+   * through all parent contexts and uses the `reducer` function to reduce the
+   * value from the parent binding chain against the last value. To avoid
+   * circular or duplicate visits, a set is used to keep track of all context
+   * objects that have been visited.
+   *
+   * @param visitor A function to produce a result from the context object
+   * locally without consulting the parents.
+   * @param reducer An optional function to reduce the value from the value
+   * from a parent binding againt the last result. If not present, a default
+   * reducer is used to use the current value to override the last one.
+   * @param visited An optional set to keep track of visited contexts.
+   */
+  protected visitAllContexts<T>(
+    visitor: (ctx: Context) => {done: boolean; value: T},
+    reducer?: (previousValue: T, currentValue: T) => T,
```

Comment:
I am wondering what is the benefit of having `visitor` and `reducer` instead of a single function performing both tasks?

The `visitor` is already performing mapping from `Context` to `T`, plus a filter condition deciding when to stop. If we want another reducer, then shouldn't we allow the reducer to convert from `T` returned by `visitor` to arbitrary type (e.g. `U`)?

Ideally, this function should not be needed and users should compose their own pipeline using low-level operations:
 - map
 - takeUntil
 - reduce

As long as these operations are lazy/pull-based, then we should have the same functionality. Below is a mock-up implementation in Javascript to show what I mean:

```js
function* map(iterable, mapper) {
  for (item of iterable) {
    yield mapper(item);
  }
}

function* takeUntil(iterable, cond) {
  for (item of iterable) {
    if (cond(item)) return;
    yield item;
  }
}

function reduce(iterable, reducer, initialValue) {
  let acc = initialValue;
  for (item of iterable) {
    acc = reducer(acc, item);
  }
  return acc;
}
```